Jack/Gen.fs

```diff
@@ -41,18 +36,69 @@ module Gen =
     let bind (m0 : Gen<'a>) (k0 : 'a -> Gen<'b>) : Gen<'b> =
         bindRandom (toRandom m0) (toRandom << k0) |> ofRandom
 
+    let apply (gf : Gen<'a -> 'b>) (gx : Gen<'a>) : Gen<'b> =
+        bind gf <| fun f ->
+        bind gx <| fun x ->
+        constant (f x)
+
+    let mapRandom (f : Random<Tree<'a>> -> Random<Tree<'b>>) (g : Gen<'a>) : Gen<'b> =
+        toRandom g |> f |> ofRandom
+
+    let mapTree (f : Tree<'a> -> Tree<'b>) (g : Gen<'a>) : Gen<'b> =
+        mapRandom (Random.map f) g
+
+    let map (f : 'a -> 'b) (g : Gen<'a>) : Gen<'b> =
+        mapTree (Tree.map f) g
+
+    let map2 (f : 'a -> 'b -> 'c) (gx : Gen<'a>) (gy : Gen<'b>) : Gen<'c> =
+        bind gx <| fun x ->
+        bind gy <| fun y ->
+        constant (f x y)
+
+    let map3 (f : 'a -> 'b -> 'c -> 'd) (gx : Gen<'a>) (gy : Gen<'b>) (gz : Gen<'c>) : Gen<'d> =
+        bind gx <| fun x ->
+        bind gy <| fun y ->
+        bind gz <| fun z ->
+        constant (f x y z)
+
+    let zip (gx : Gen<'a>) (gy : Gen<'b>) : Gen<'a * 'b> =
+        map2 (fun x y -> x, y) gx gy
+
+    let zip3 (gx : Gen<'a>) (gy : Gen<'b>) (gz : Gen<'c>) : Gen<'a * 'b * 'c> =
+        map3 (fun x y z -> x, y, z) gx gy gz
+
     type Builder internal () =
         member __.Return(a) =
             constant a
         member __.ReturnFrom(g) =
             g
         member __.Bind(m, k) =
             bind m k
-        member __.Zero() =
-            constant ()
 
     let private gen = Builder ()
 
+    //
+    // Combinators - Shrinking
+    //
+
+    /// Prevent a 'Gen' from shrinking.
+    let noShrink (g : Gen<'a>) : Gen<'a> =
+        let drop (Node (x, _)) =
+            Node (x, LazyList.empty)
+        mapTree drop g
+
+    /// Apply an additional shrinker to all generated trees.
+    let shrinkLazy (f : 'a -> LazyList<'a>) (g : Gen<'a>) : Gen<'a> =
+        mapTree (Tree.expand f) g
+
+    /// Apply an additional shrinker to all generated trees.
+    let shrink (f : 'a -> List<'a>) (g : Gen<'a>) : Gen<'a>  =
+        shrinkLazy (LazyList.ofList << f) g
```

Comment:
This combinator is pretty neat I think 😄 You can apply an extra shrinking function to an existing generator - there's an example in the script.